lib/rouge/lexers/csound.rb

```diff
@@ -0,0 +1,503 @@
+        rule /\bopcode\b/, Keyword::Declaration, :after_opcode_keyword
+        rule /\b(?:end(?:in|op))\b/, Keyword::Declaration
+
+        mixin :partial_statements
+      end
+
+      state :partial_statements do
+        rule /\b(?:0dbfs|A4|k(?:r|smps)|nchnls(?:_i)?|sr)\b/, Name::Variable::Global
+
+        mixin :numbers
+
+        rule %r(\+=|-=|\*=|/=|<<|>>|<=|>=|==|!=|&&|\|\||[~¬]|[=!+\-*/^%&|<>#?:]), Operator
+        rule /[(),\[\]]/, Punctuation
+
+        rule /"/, Str, :quoted_string
+        rule /{{/, Str, :braced_string
+
+        rule /\b(?:do|else(?:if)?|end(?:if|until)|fi|i(?:f|then)|kthen|od|then|until|while)\b/, Keyword
+        rule /\br(?:ir)?eturn\b/, Keyword::Pseudo
+
+        rule /\b[ik]?goto\b/, Keyword, :goto_before_label
+        rule /\b(r(?:einit|igoto)|tigoto)(\(|\b)/ do
+          groups Keyword::Pseudo, Punctuation
+          push :goto_before_label
+        end
+        rule /\b(c(?:g|in?|k|nk?)goto)(\(|\b)/ do
+          groups Keyword::Pseudo, Punctuation
+          push :goto_before_label
+          push :goto_before_argument
+        end
+        rule /\b(timout)(\(|\b)/ do
+          groups Keyword::Pseudo, Punctuation
+          push :goto_before_label
+          push :goto_before_argument
+          push :goto_before_argument
+        end
+        rule /\b(loop_[gl][et])(\(|\b)/ do
+          groups Keyword::Pseudo, Punctuation
+          push :goto_before_label
+          push :goto_before_argument
+          push :goto_before_argument
+          push :goto_before_argument
+        end
+
+        rule /\bprintk?s\b/, Name::Builtin, :prints_opcode
+        rule /\b(?:readscore|scoreline(?:_i)?)\b/, Name::Builtin, :csound_score_opcode
+        rule /\bpyl?run[it]?\b/, Name::Builtin, :python_opcode
+        rule /\blua_(?:exec|opdef)\b/, Name::Builtin, :lua_opcode
+        rule /\bp\d+\b/, Name::Variable::Instance
+        rule /\b(#{Csound.identifier})(?:(:)([A-Za-z]))?\b/ do |m|
+          name = m[1]
+          if self.class.opcodes.include?(name) || self.class.deprecated_opcodes.include?(name)
+            token Name::Builtin, name
+            if m[2]
+              token Punctuation, m[2]
+              token Keyword::Type, m[3]
+            end
+          elsif user_defined_opcodes.include? name
+            token Name::Function, name
+          else
+            if m = /^(g?[aikSw])(\w+)/.match(name)
+              token Keyword::Type, m[1]
+              token Name, m[2]
+            else
+              token Name, name
+            end
+          end
+        end
+      end
+
+      state :instrument_numbers_and_identifiers do
+        mixin :whitespace_and_macro_uses
+        rule /\d+|#{Csound.identifier}/, Name::Function
+        rule /[+,]/, Punctuation
+        rule /\n/, Text, :pop!
+      end
+
+      state :after_opcode_keyword do
+        mixin :whitespace_and_macro_uses
+        rule Csound.identifier do |m|
+          user_defined_opcodes.add m[0]
```

Comment:
I'd like to avoid this kind of statefulness if at all possible. We aren't interpreting the language, just highlighting it ;)